module/wrappers.py

```diff
@@ -13,20 +13,26 @@
 
 def IDWrapper(freer):
     """
-    Classes create with IDWrapper return an ID and then free it upon exit of the
-    context.
+    Classes create with IDWrapper return an ID and then free it upon exit of
+    the context or when garbage collected, whichever comes first.
     """
     class Wrapper:
         def __init__(self, conn):
-            self.conn = conn
-            self.id = None
+            super().__init__()
+            self.id = conn.generate_id()
+            self._free = lambda self=self: getattr(conn.core, freer)(self.id)
```

Comment:
Is there any reason this can't be an actual member function?

Reply:
No